cpp/src/reductions/reductions.cpp

```diff
@@ -17,70 +17,91 @@
 #include <cudf/reduction.hpp>
 #include <cudf/scalar/scalar_factories.hpp>
 #include <cudf/detail/reduction_functions.hpp>
+#include <cudf/detail/aggregation/aggregation.hpp>
 
 namespace cudf {
 namespace experimental {
 namespace detail {
 
-std::unique_ptr<scalar> reduce(
-    column_view const& col, reduction_op op, data_type output_dtype,
-    size_type ddof, 
-    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
-    cudaStream_t stream = 0)
-{
-  std::unique_ptr<scalar> result = make_default_constructed_scalar(output_dtype);
-  result->set_valid(false, stream);
+struct reduce_dispatch_functor {
+  column_view const col;
+  data_type output_dtype;
+  rmm::mr::device_memory_resource *mr;
+  cudaStream_t stream;
 
-  // check if input column is empty
-  if (col.size() <= col.null_count()) return result;
+  reduce_dispatch_functor(column_view const &col, data_type output_dtype,
+                          rmm::mr::device_memory_resource *mr,
+                          cudaStream_t stream)
+      : col(col), output_dtype(output_dtype), mr(mr), stream(stream) {}
 
-  switch (op) {
-    case reduction_op::SUM:
-      result = reduction::sum(col, output_dtype, mr, stream);
+  template <aggregation::Kind k>
+  std::unique_ptr<scalar> operator()(std::unique_ptr<aggregation> const &agg) {
+    switch (k) {
```

Comment:
There is also a new aggregation dispatcher added as part of the infra. Would require you to restructure the `reduction::<agg>` as template specializations though.

Reply:
I am using `dispatch_aggregation`. I am using this switch statement to keep it more readable. The switch statement arg is a template parameter, so essentially each case statement block is specialization code.